doc: add versions information to the contribution guidelines (#1528)
kubernetes-sigs/kubebuilder: CONTRIBUTING.md

Check the CI job after to do the Pull Request and then, click on in the `Details` of `netlify/kubebuilder/deploy-preview`

## Understanding the versions

|   Name	|   Example	|  Description |
|---	|---	|---	|

> |  PROJECT version |  `V1`,`V2`,`V3` | After the Kubebuilder v3 introduces the plugins architecture to [Extensible CLI and Scaffolding Plugins](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/extensible-cli-and-scaffolding-plugins-phase-1.md), the PROJECT version represents the version of the PROJECT file layout which are generated in the projects built with it.  |

Reviewer: ```suggestion
|  PROJECT version |  `V1`,`V2`,`V3` | As of the introduction of [Extensible CLI and Scaffolding Plugins](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/extensible-cli-and-scaffolding-plugins-phase-1.md), PROJECT version represents the layout of PROJECT file itself.  For `v1` and `v2` projects, there's extra meaning -- see below.  |
```

Author: Done

> |  Releases version | `v2.2.0`, `v2.3.0`, `v2.3.1` | Represents the release tags of this project. See the [releases](https://github.com/kubernetes-sigs/kubebuilder/releases) page. |

Reviewer: ```suggestion
|  Release version | `v2.2.0`, `v2.3.0`, `v2.3.1` | Tagged versions of the KubeBuilder project, representing changes to the source code in this repository. See the [releases](https://github.com/kubernetes-sigs/kubebuilder/releases) page. |
```

> |  Plugin Versions | `go.kubebuilder.io/v2.0.0` | Represents the code source of the Plugin used to scaffold the projects. |

Reviewer: ```suggestion
|  Plugin Versions | `go.kubebuilder.io/v2.0.0` | Represents the version of an individual plugin, as well as the corresponding scaffolding that it generates. |
```

Author: Done

> Note that, we just need to bump a new PROJECT version if a breaking or significant change be introduce in the PROJECT file. Otherwise, it not requires be changed. And then, see that Releases and Plugin versions following the [semver](https://semver.org/) definition.  

Reviewer: ```suggestion
Note that PROJECT version should only be bumped if a breaking change is introduced in the PROJECT file format itself.  Changes to the Go scaffolding or the KubeBuilder CLI *do not* affect the PROJECT version.

Similarly, the introduction of a new major version (`(X+1).0.0`) of the Go plugin might only lead to a new minor (`X.(Y+1).0`) release of KubeBuilder, since no breaking change is being made to the CLI itself.  It'd only be a breaking change to KubeBuilder if we remove support for an older version of the plugin.

For more information on how the release and plugin versions work, see the the [semver](https://semver.org/) documentation.
```

Reviewer: ```markdown
Note that in the case of the `v1` and `v2` PROJECT version, a corresponding plugin version is implied and constant -- `v1` implies the `go.kubebuilder.io/v1` plugin, and similarly for `v2`.  This is for legacy purposes -- no such implication is made with the `v3` PROJECT version.
```

Author: Done:

## Introducing changes in the scaffold files

> Changes in the scaffold files require a new Plugin version. Note that, if we delete and/or update a file or a content that is scaffolded by default, it will result in a breaking change and requires a `MAJOR` Plugin version. But, if we are adding a new content and/or file it probably will not be and breaking change. 
> 
> For our sake let's try to simplify it and consider that any change that will break the expected behaviour of a project built with the previous `MINOR` Plugin versions will be a breaking change. 

Reviewer: ```suggestion
Changes in the scaffolded files require a new Plugin version. If we delete or update a file that is scaffolded by default, it's a breaking change and requires a `MAJOR` Plugin version.  If we add a new file, it may not be a breaking change.

More simply: any change that will break the expected behaviour of a project built with the previous `MINOR` Plugin versions is a breaking change to that plugin. 
```

Author: Done

**Example**    

> The master version has been scaffold projects with the plugin version `2.5.1` by default then, if the changes introduced in our PR will not work well with the projects which were built with the plugin versions >= `2.0.0` <= `2.5.1` without we perform manual changes in these projects then, it means that; it will be a breaking change and requires a `MAJOR` Plugin version. 
> 
> So, in this scenario, we should add to the PR a guide with the required steps that need to be done by the users to upgrade their projects built with `2.X.X` in order to make them work with the new `MAJOR` Plugin version `3.0.0`.
> 
> **IMPORTANT** The breaking changes cannot be applied in the V2 projects. So, beware and ensure it.
> 

Reviewer: ```suggestion
The current latest KubeBuilder version (`5.3.1`) scaffolds projects with the plugin version `3.2.1` by default.

The changes introduced in our PR will not work well with the projects which were built with the plugin versions `3.0.0...3.5.1` without users taking manual steps to update their projects.

Thus, our changes introduce a breaking change to the Go plugin, and require a `MAJOR` Plugin version bump.

In the PR, we should add a migration guide to the [Migrations](https://book.kubebuilder.io/migrations.html) section of the KubeBuilder book.  It should detail the required steps that users should take to upgrade their projects from `go.kubebuilder.io/3.X.X` to the new `MAJOR` Plugin version `go.kubebuilder.io/4.0.0`.

This also means we should introduce a new KubeBuilder minor version `5.4.0`: since we've only added a new plugin version without removing the old one, this is considered a new feature to KubeBuilder, and not a breaking change.

**IMPORTANT** Breaking changes cannot be made to PROJECT versions v1 and v2, and consequently plugin versions `go.kubebuilder.io/1` and `go.kubebuilder.io/2`.
```

Author: I did small changes in the suggestion which I think make easier to understand it. 
- KubeBuilder version (`5.3.1`)  -> KubeBuilder **Release** version (`5.3.1`)
- This also means we should introduce a new KubeBuilder minor version `5.4.0`: -> This also means we should introduce a new KubeBuilder minor version `5.4.0` **when the project be released:**

Author: done

Similarly, the introduction of a new major version (`(X+1).0.0`) of the Go plugin might only lead to a new minor (`X.(Y+1).0`) release of KubeBuilder, since no breaking change is being made to the CLI itself.  It'd only be a breaking change to KubeBuilder if we remove support for an older version of the plugin.

For more information on how the release and plugin versions work, see the the [semver](https://semver.org/) documentation.

**NOTE:** In the case of the `v1` and `v2` PROJECT version, a corresponding Plugin version is implied and constant -- `v1` implies the `go.kubebuilder.io/v1` Plugin, and similarly for `v2`.  This is for legacy purposes -- no such implication is made with the `v3` PROJECT version.

## Introducing changes in the scaffold files

Changes in the scaffolded files require a new Plugin version. If we delete or update a file that is scaffolded by default, it's a breaking change and requires a `MAJOR` Plugin version.  If we add a new file, it may not be a breaking change.

**More simply:** any change that will break the expected behaviour of a project built with the previous `MINOR` Plugin versions is a breaking change to that plugin. 

**EXAMPLE:**

> The current latest KubeBuilder Release version (`5.3.1`) scaffolds projects with the plugin version `3.2.1` by default.

Author: @DirectXMan12 
I did small changes in the suggestion which I think make easier to understand it.

KubeBuilder version (5.3.1) -> KubeBuilder Release version (5.3.1)
This also means we should introduce a new KubeBuilder minor version 5.4.0: -> This also means we should introduce a new KubeBuilder minor version 5.4.0 when the project be released:

The changes introduced in our PR will not work well with the projects which were built with the plugin versions `3.0.0...3.5.1` without users taking manual steps to update their projects.

> Thus, our changes introduce a breaking change to the Go plugin, and require a `MAJOR` Plugin version bump.

Reviewer 2: Should join this with the paragraph above for continuity.
